src/main/buffered_data/batched_buffered_data.cpp

```diff
@@ -0,0 +1,220 @@
+	// printf("current-batch tuple_count: %llu\n", current_batch_tuple_count.load());
+	// printf("other-batches tuple_count: %llu\n\n", other_batches_tuple_count.load());
+
+	stack<idx_t> to_remove;
+	for (auto &it : in_progress_batches) {
+		auto batch = it.first;
+		auto &buffered_chunks = it.second;
+		if (batch > min_batch) {
+			// This batch is still in progress, can not be fetched from yet
+			break;
+		}
+		if (batch != min_batch && !buffered_chunks.completed) {
+			// We haven't completed this batch yet
+			break;
+		}
+		D_ASSERT(buffered_chunks.completed || batch == min_batch);
+
+		// We have already materialized chunks, have to move them to `batches` so they can be scanned
+		auto &chunks = buffered_chunks.chunks;
+
+		idx_t tuple_count = 0;
+		for (auto it = chunks.begin(); it != chunks.end(); it++) {
+			auto chunk = std::move(*it);
+			tuple_count += chunk->size();
+			batches.push_back(std::move(chunk));
+		}
+		other_batches_tuple_count -= tuple_count;
+		current_batch_tuple_count += tuple_count;
+		to_remove.push(batch);
+	}
+	while (!to_remove.empty()) {
+		auto batch = to_remove.top();
+		to_remove.pop();
+		in_progress_batches.erase(batch);
+	}
+}
+
+PendingExecutionResult BatchedBufferedData::ReplenishBuffer(StreamQueryResult &result,
+                                                            ClientContextLock &context_lock) {
+	if (Closed()) {
+		return PendingExecutionResult::EXECUTION_ERROR;
+	}
+	if (BufferIsFull()) {
+		// The buffer isn't empty yet, just return
+		return PendingExecutionResult::RESULT_READY;
+	}
+	UnblockSinks();
+	// Let the executor run until the buffer is no longer empty
+	auto cc = context.lock();
+	auto res = cc->ExecuteTaskInternal(context_lock, result);
+	while (!PendingQueryResult::IsFinished(res)) {
+		if (BufferIsFull()) {
+			break;
+		}
+		// Check if we need to unblock more sinks to reach the buffer size
+		UnblockSinks();
+		res = cc->ExecuteTaskInternal(context_lock, result);
+	}
+	// printf("IS_FINISHED\n");
+	return res;
+}
+
+void BatchedBufferedData::CompleteBatch(idx_t batch) {
+	auto it = in_progress_batches.find(batch);
+	if (it == in_progress_batches.end()) {
+		return;
+	}
+
+	auto &buffered_chunks = it->second;
+	buffered_chunks.completed = true;
+}
+
+unique_ptr<DataChunk> BatchedBufferedData::Scan() {
+	unique_ptr<DataChunk> chunk;
+	bool empty = false;
+	{
+		lock_guard<mutex> lock(glock);
+		if (!batches.empty()) {
+			chunk = std::move(batches.front());
+			batches.pop_front();
+		} else {
+			empty = true;
+		}
+	}
+	if (empty) {
+		D_ASSERT(!chunk);
+		// Increase the min batch index to see if there are still completed batches
+		// waiting to be processed
+		auto it = in_progress_batches.begin();
+		if (it != in_progress_batches.end()) {
+			auto batch = it->first;
+			auto &chunks = it->second;
+			if (chunks.completed) {
+				// By updating the min batch index, we'll move the chunks to 'batches';
+				UpdateMinBatchIndex(batch);
```

Comment:
Even worse, we call UpdateMinBatchIndex, which can remove items from `in_progress_batches`, invalidating the iterators